Feat/e2e list production

slebastard opened the pull request:
On the refactoring side, this PR:
- applies some minor refactoring on model configuration for model `small`
- moves method `evaluate` from `__main__.py` to module `evaluate.py`
- changes the names of model run folders to that model name is used, for easier retrieval
- introduces new CLI option `save_model` to save each component of the model pipeline as pickle into `model_comp{comp_id}.pickle` 
- introduces CLI option `predict_siret_list` to specify a filepath containing a list of SIRETs. Dataset `predict`is then restricted to SIRETs in this list
- introduces CLI option `explain` to specify whether the fields containing the explanations of model predictions should be computed. This option has a store_true behaviour, such that by default, explanations fields are **not** computed. 

Moreover, the PR introduces tools to select model thresholds from predicted failure risk, allowing to build lists of selected companies using a few intuitive functions:
- introduces methods to select thresholds from various criteria in `evaluate.py`
- introduces notebook ` 03-select-thresholds.ipynb` to leverage these new functions
- introduces notebook `04-merge_models_into_list.ipynb` to use the selected thresholds to produce a list

slebastard committed:
feat: updating small model configuration to initiate datasets

slebastard committed:
feat: adding CLI option to predict on a given list of SIRETs

slebastard committed:
fix: correcting for misspelled field in ARGS_TO_ATTRS dict

slebastard committed:
feat: updating model_runs save folder and file names

slebastard committed:
feat: adding functions to merge a list of predictions and produce lists

slebastard committed:
feat: Added option to save model parameters to pickle

slebastard committed:
feat: removing functions to merge models from notebook 02

slebastard committed:
feat: moving evaluation functions into module evaluate.py

slebastard committed:
feat: adding CLI option explain to compute and export explanations fields

slebastard committed:
Adding AUCPR metric to evaluation notebook (03)

slebastard committed:
feat: updating field names for export of Score to Mongo

slebastard committed:
feat: Moving list production functions to make_list.py + Adding threshold selection

1. Moved functions `merge_models`, `assign_flag` and `make_alert` from `model_selection.py` to `make_list.py`
2. Added functions `make_precision_recall_curve`, `make_thresholds_from_fbeta`, `make_thresholds_from_conditions` to help selection of list thresholds

slebastard committed:
feat: adding package tqdm to `requirement.txt`

`tqdm` is used by function `make_thresholds_from_fbeta` in module `evaluate`

slebastard committed:
fix: replacing np.arange with np.linspace in thr sel by f_beta

slebastard committed:
feat: adding notebook 03-select-thresholds to tune thresholds for list prod

slebastard committed:
refacto: removing cells for using local functions in notebooks/03

slebastard committed:
refacto: renaming evaluation & list production notebooks

vviers commented on models/small/model_conf.py:
```diff
@@ -92,15 +93,34 @@
 PREDICT_SAMPLE_SIZE = 1_000_000_000 if ENV == "prod" else 5_000
+PREDICT_DATASET = SFDataset(
+    date_min=PREDICT_ON,
+    date_max=PREDICT_ON[:-2] + "28",
```
Some day we'll find a cleaner way to do this 😆

vviers commented on predictsignauxfaibles/__main__.py:
```diff
@@ -54,6 +56,10 @@ def get_train_test_predict_datasets(args_ns: argparse.Namespace, conf: ModuleTyp
     }
 
+    if args_ns.predict_siret_list is not None:
+        predict_siret_list = pd.read_csv(args_ns.predict_siret_list).siret.tolist()
```
So the file needs to contain a `siret` column ? Could we be more flexible here (e.g. a .txt file with one siret per line) ?

The idea is that the siret file will often be produced by another process that does not necessarily split csv with a column name :

```sh
./my_siret_generating_process.sh > sirets.txt
python -m predictsignauxfaibles --predict_siret_list sirets.txt
```

Granted, it's pretty easy to reformat such file, but the expected format should be in the help menu

Finally, you might wan't to raise an informative exception here in case you get an `AttributeError: 'DataFrame' object has no attribute 'siret'`

slebastard replied:
Now using the following code for reading siret list from argument `predict_siretlist_path` if provided:
```
predict_siret_list = (
            pd.read_csv(
                args_ns.predict_siretlist_path,
                names=["siret"],
                header=0,
                index_col=False,
            )
            .siret.astype(str)
            .tolist()
        )
```
This is largely fault-tolerant: if there are more than one columns, as long as SIRETs are in the first columns are columns are comma-separated, SIRETS will be loaded successfully.

vviers commented on predictsignauxfaibles/__main__.py:
```diff
@@ -258,6 +264,12 @@ def make_parser():
+        "--predict_siret_list",
+        type=str,
+        dest="predict_siret_list",
+        help="If provided a csv list of SIRETs, will predict on these specific SIRET",
```
Make expected format explicit here :)

slebastard replied:
We now specify in the CLI helper that:
- no header should be provided
- no index other than SIRETs should be provided

Moreover, reading an empty file will result in a `EmptyInputError`

See commit `d05655c9b190b6f8ea168ca9f4540d050150f91b`

vviers commented on predictsignauxfaibles/model_selection.py:
```diff
@@ -50,3 +53,66 @@ def make_sf_train_test_splits(
+
+def merge_models(model_list: list):
+    # pylint: disable=all
+    # As pylint does not seem to handle recursions properly
```
maybe you can only disable the specific pylint error instead of all ? Or is it a bigger problem ?

slebastard replied:
It looks like pylint fails when inspecting function `merge_models` due to some `RecursionError` that I have yet to investigate in detail:
`RecursionError: maximum recursion depth exceeded while calling a Python object`
Before we investigate that in-depth, disabling pylint checks for function `merge_models`

The following issue seems to point to operations on pandas DataFrame as a source of recursion error from `pylint`, but no solution was offered yet:
https://github.com/volkamerlab/opencadd/issues/63
Unfortunately, pylint has removed in the v2.0.0 an option `--optimize-ast`that provided a clean workaround. The only remaining workaround seems to be to downgrade package `astroid` (see https://github.com/PyCQA/pylint/issues/3982#issuecomment-745236956). If we have time, maybe we couls comment on this last issue...

In the meantime, we are stuck with having function `merge_models` uncovered by linting.
Because `# pylint: disable=all` disables pylint for the entire file rather than for a single function, let's leave `merge_models` as the only function in a module `predictsignauxfaibles.merge_models.py`, with `pylint` disabled.

vviers commented on predictsignauxfaibles/model_selection.py:
```diff
@@ -50,3 +53,66 @@ def make_sf_train_test_splits(
     return out
+
+
+def merge_models(model_list: list):
```
I think this type hint would be clearer :

```
from typing import List
import pandas as pd

def merge_models(model_list: List[pd.DataFrame]):
    ...
```

slebastard replied:
Ah yes, I wasn't aware that this kind of type hints were possible, thanks!
Modified in commit `2b1f361b7dfc6a999c06188c3dbcf13855887ce2`

vviers commented on predictsignauxfaibles/model_selection.py:
```diff
@@ -50,3 +53,66 @@ def make_sf_train_test_splits(
+    """
+    merged = model_list.pop()
+    for model_id in range(len(model_list)):
+        model = model_list.pop()
```
Since you popped the first df (great idea), this can be simplified as : 

```python
for model in model_list:
    merged = pd.merge(...)
```

slebastard replied:
Yep, that's cleaner. Fixed in commit `2b1f361b7dfc6a999c06188c3dbcf13855887ce2`

vviers commented on predictsignauxfaibles/model_selection.py:
```diff
@@ -50,3 +53,66 @@ def make_sf_train_test_splits(
+    return merged
+
+
+def assign_flag(pred: float, t_rouge: float, t_orange: float):
```
I would check that t_rouge > t_orange and raise an error if not

slebastard replied:
Added assertions to verify that arguments `t_rouge` and `t_orange` to function `assign_flag`:
    - are floats between 0 and 1
    - verify `t_rouge >= t_orange`
    In particular, we allow `t_rouge==t_orange` as it enables to have a single, unique threshold
    See commit `f19e89b6c1d415e78e5cb29dccb28c6e17a624b3`

vviers commented on predictsignauxfaibles/model_selection.py:
```diff
@@ -50,3 +53,66 @@ def make_sf_train_test_splits(
+    return "Pas d'alerte"
+
+
+def split_predictions(preds: pd.DataFrame, t_rouge: float, t_orange: float):
```
I think we can come up with a more informative function name
`test_splits` ? `log_splits_size` ?

slebastard replied:
Actually I don't like the fact that function `make_alert` both adds an `alert` columns to its `pd.DataFrame` input `preds`, **and** logs list sizes.
 I split these two steps:
    - function `make_list.log_splits_size` only logs potential list sizes, given two thresholds. This allows to check the size of list, as a help for threshold selection
    - function `make_list.assign_flag` should be applied on a `pd.DataFrame` row `predicted_probability` to actually produce field `alert`

See use of both functions in `notebooks/04-merge_models_into_list.ipynb`

See commit `e191434a41bf140211f233f7d6788cd6b78e5e00`

vviers commented on predictsignauxfaibles/make_list.py:
```diff
@@ -0,0 +1,141 @@
+# pylint: disable=all
```
I don't think disabling pylint for the whole module if a good idea 😕

vviers replied:
Also I think these functions could go in `utils.py` but I'm not against having them in a separate module :)

slebastard replied:
See comment above:

> It looks like pylint fails when inspecting function merge_models due to some RecursionError that I have yet to investigate in detail:
> RecursionError: maximum recursion depth exceeded while calling a Python object
> Before we investigate that in-depth, disabling pylint checks for function merge_models
> 
> The following issue seems to point to operations on pandas DataFrame as a source of recursion error from pylint, but no solution was offered yet:
> volkamerlab/opencadd#63
> Unfortunately, pylint has removed in the v2.0.0 an option --optimize-astthat provided a clean workaround. The only remaining workaround seems to be to downgrade package astroid (see PyCQA/pylint#3982 (comment)). If we have time, maybe we couls comment on this last issue...
> 
> In the meantime, we are stuck with having function merge_models uncovered by linting.
> Because # pylint: disable=all disables pylint for the entire file rather than for a single function, let's leave merge_models as the only function in a module predictsignauxfaibles.merge_models.py, with pylint disabled.

See commit `66518bb2038d28960c47eecf98c809418bc01808`

vviers reviewed (commented):
Thanks for this great work ! 💯 

A couple of comments to resolve/answer and we should be good to merge :)

slebastard committed:
feat: introduced CLIError and derived for handling bad CLI inputs

- introduced base class CLIError and inherited EmptyInputError for handling erroneous CLI inputs
- detailed help message for option `predict_siret_list`.
Any file can be taken as input, and will be parsed as CSV.
There should be no header, and SIRET should be contained in the first column if there are multiple columns.

slebastard committed:
refacto: make_list.merge_models

slebastard committed:
feat: asserting validity of inputs to `make_list.assign_flag`

Ading checks to verify that arguments `t_rouge` and `t_orange` to function `assign_flag`:
- are floats between 0 and 1
- verify t_rouge >= t_orange
In particular, we allow `t_rouge==t_orange` as it enables to have a single, unique threshold

slebastard committed:
refacto: `make_list.make_alert`-->`make_list.log_splits_size`

Function `make_alert` used to both add an `alert` columns to its `pd.DataFrame` input `preds`, and log list sizes.
These two functionnalities are now separated:
- function `make_list.log_splits_size` only logs potential list sizes, given two thresholds. This allows to check the size of list, as a help for threshold selection
- function `make_list.assign_flag` should be applied on a `pd.DataFrame` row `predicted_probability` to actually produce field `alert`

See use of both functions in `notebooks/04-merge_models_into_list.ipynb`

slebastard committed:
refacto: moving `merge_models` into module `merge_models.py`

Because `pylint` currently fails to handle some operations on `pd.DataFrame` being performed in function `merge_models`,
and because disabling pylint is only possible module-wide, moving function merge_models into its own module `predictsignauxfaibles.merge_models`

Moreover:
- moving other functions from `predictsignauxfaibles.make_list` to `predictsignauxfaibles.utils`
- creating doctype for function `utils.assign_flag`
- correcting for function names imported from `utils` in `notebooks/04-merge_models_into_list.ipynb`

vviers commented on predictsignauxfaibles/utils.py:
```diff
@@ -2,8 +2,10 @@
 import importlib.util
 import logging
 import math
```
Unused import ?

vviers reviewed (approved):
LGTM ! Thanks ! 💯
